File: platform/polyglot/github_api/py/github_api/middleware/github_hooks.py
"""
Response processing hooks for GitHub API responses.

These hooks extract metadata from httpx.Response objects for use
by the SDK client and middleware layers.
"""

from __future__ import annotations

from typing import Any

import httpx

from github_api.sdk.rate_limit import RateLimitInfo, parse_rate_limit_headers

__all__ = [
    "response_204_hook",
    "json_fallback_hook",
    "request_id_hook",
    "rate_limit_hook",
]


def response_204_hook(response: httpx.Response) -> dict[str, Any]:
    """Handle 204 No Content responses by returning an empty dict.

    Args:
        response: The httpx response object.

    Returns:
        Empty dict for 204 responses, or None for other status codes.
    """
    if response.status_code == 204:
        return {}
    return {}


def json_fallback_hook(response: httpx.Response) -> dict[str, Any]:
    """Safely parse JSON from a response, falling back to text wrapper.

    Args:
        response: The httpx response object.

    Returns:
        Parsed JSON dict, or {"data": text} on parse failure.
    """
    try:
        data = response.json()
        if isinstance(data, dict):
            return data
        return {"data": data}
    except Exception:
        text = response.text
        return {"data": text} if text else {}


def request_id_hook(response: httpx.Response) -> str | None:
    """Extract the x-github-request-id header from a response.

    Args:
        response: The httpx response object.

    Returns:
        The request ID string, or None if not present.
    """
    return response.headers.get("x-github-request-id")


def rate_limit_hook(response: httpx.Response) -> RateLimitInfo | None:
    """Extract rate limit information from response headers.

    Args:
        response: The httpx response object.

    Returns:
        RateLimitInfo if rate limit headers are present, None otherwise.
    """
    return parse_rate_limit_headers(response.headers)

File: platform/polyglot/github_api/mjs/src/middleware/github-hooks.mjs
/**
 * Response processing hooks for GitHub API responses.
 * These hooks normalize and extract metadata from API responses.
 * @module middleware/github-hooks
 */

import { parseRateLimitHeaders } from '../sdk/rate-limit.mjs';

/**
 * Normalize a 204 No Content response to an empty object.
 * @param {Response} response - Fetch Response object
 * @returns {Object|null} Empty object for 204, null otherwise
 */
export function response204Hook(response) {
  if (response.status === 204) {
    return {};
  }
  return null;
}

/**
 * Fallback for JSON parse failures: returns the raw text wrapped in an object.
 * @param {Response} response - Fetch Response object
 * @returns {Promise<Object|null>} Object with data property, or null if JSON parsed fine
 */
export async function jsonFallbackHook(response) {
  const contentType = response.headers.get('content-type') || '';
  if (!contentType.includes('application/json')) {
    const text = await response.text();
    return { data: text };
  }
  return null;
}

/**
 * Extract x-github-request-id from response headers.
 * @param {Response} response - Fetch Response object
 * @returns {string|undefined} The request ID if present
 */
export function requestIdHook(response) {
  return response.headers.get('x-github-request-id') || undefined;
}

/**
 * Parse rate limit headers from a response.
 * @param {Response} response - Fetch Response object
 * @returns {import('../sdk/rate-limit.mjs').RateLimitInfo|null} Parsed rate limit info
 */
export function rateLimitHook(response) {
  return parseRateLimitHeaders(response.headers);
}
